File: backend/VocabuAI.Api/Infrastructure/AppBuildInfo.cs
using System.Globalization;
using System.Reflection;

namespace VocabuAI.Api.Infrastructure;

public sealed record AppBuildInfo(
    string ApplicationName,
    string Version,
    string? CommitSha,
    string? Branch,
    DateTimeOffset? BuildTimeUtc)
{
    public static AppBuildInfo Create(string applicationName)
    {
        var assembly = Assembly.GetExecutingAssembly();
        var version = Environment.GetEnvironmentVariable("APP_VERSION")
            ?? assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion
            ?? assembly.GetName().Version?.ToString()
            ?? "0.2.2";

        var commitSha = Normalize(Environment.GetEnvironmentVariable("APP_COMMIT_SHA"));
        var branch = Normalize(Environment.GetEnvironmentVariable("APP_BRANCH"));
        var buildTimeRaw = Normalize(Environment.GetEnvironmentVariable("APP_BUILD_TIME_UTC"));

        DateTimeOffset? buildTimeUtc = null;
        if (!string.IsNullOrWhiteSpace(buildTimeRaw) &&
            DateTimeOffset.TryParse(
                buildTimeRaw,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                out var parsed))
        {
            buildTimeUtc = parsed;
        }

        return new AppBuildInfo(
            applicationName,
            version,
            commitSha,
            branch,
            buildTimeUtc);
    }

    private static string? Normalize(string? value)
        => string.IsNullOrWhiteSpace(value) ? null : value.Trim();
}

File: frontend/src/infrastructure/appBuildInfo.ts
import Constants from "expo-constants";
import type { AppVersionDto } from "../domain/dtos/AppVersionDto";

export function getFrontendBuildInfo(): AppVersionDto {
  const extra = Constants.expoConfig?.extra ?? {};

  return {
    applicationName: "Frontend",
    version:
      typeof extra.appVersion === "string" && extra.appVersion.trim()
        ? extra.appVersion.trim()
        : "0.2.2",
    commitSha:
      typeof extra.appCommitSha === "string" && extra.appCommitSha.trim()
        ? extra.appCommitSha.trim()
        : null,
    branch:
      typeof extra.appBranch === "string" && extra.appBranch.trim()
        ? extra.appBranch.trim()
        : null,
    buildTimeUtc:
      typeof extra.appBuildTimeUtc === "string" && extra.appBuildTimeUtc.trim()
        ? extra.appBuildTimeUtc.trim()
        : null
  };
}
